nexus/src/app/crucible.rs

```diff
@@ -0,0 +1,1051 @@
+
+        match result {
+            Ok(v) => Ok(Some(v.into_inner())),
+
+            Err(e) => {
+                error!(
+                    log,
+                    "region_get saw {:?}",
+                    e;
+                    "region_id" => %region_id,
+                    "dataset_id" => %dataset_id,
+                );
+
+                match e {
+                    ProgenitorOperationRetryError::Gone => {
+                        // Return an error if Nexus is unable to query the
+                        // dataset's agent for the requested region 
+                        Err(Error::Gone)
+                    }
+
+                    ProgenitorOperationRetryError::GoneCheckError(_) => {
+                        Err(Error::internal_error(
+                            "insufficient permission for crucible_agent_gone_check"
+                        ))
+                    }
+
+                    ProgenitorOperationRetryError::ProgenitorError(e) => match e {
+                        crucible_agent_client::Error::ErrorResponse(rv) => {
+                            match rv.status() {
+                                http::StatusCode::NOT_FOUND => {
+                                    Ok(None)
+                                }
+
+                                status if status.is_client_error() => {
+                                    Err(Error::invalid_request(&rv.message))
+                                }
+
+                                _ => {
+                                    Err(Error::internal_error(&rv.message))
+                                }
+                            }
+                        }
+
+                        _ => {
+                            Err(Error::internal_error(
+                                "unexpected failure during `region_get`",
+                            ))
+                        }
+                    }
+                }
+            }
+        }
+    }
+
+    async fn get_crucible_region_snapshots(
+        &self,
+        log: &Logger,
+        dataset: &db::model::Dataset,
+        region_id: Uuid,
+    ) -> Result<GetSnapshotResponse, Error> {
+        let client = self.crucible_agent_client_for_dataset(dataset);
+        let dataset_id = dataset.id();
+
+        let result = ProgenitorOperationRetry::new(
+            || async {
+                client
+                    .region_get_snapshots(&RegionId(region_id.to_string()))
+                    .await
+            },
+            || async { self.crucible_agent_gone_check(dataset_id).await },
+        )
+        .run(log)
+        .await;
+
+        match result {
+            Ok(v) => Ok(v.into_inner()),
+
+            Err(e) => {
+                error!(
+                    log,
+                    "region_get_snapshots saw {:?}",
+                    e;
+                    "region_id" => %region_id,
+                    "dataset_id" => %dataset_id,
+                );
+
+                match e {
```

Comment:
After seeing the third one of these, the bee in my bonnet is now a hornet with an anger issue. Is it possible to wrap these checks into a function and reuse it?

Reply:
I hacked on this for longer than I thought before settling on something that I liked! See ff3915033